File: app/src/main/java/net/xzos/UpgradeAll/ui/viewmodels/adapters/LogItemadapter.java
package net.xzos.UpgradeAll.ui.viewmodels.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import net.xzos.UpgradeAll.R;
import net.xzos.UpgradeAll.ui.viewmodels.LogItemView;

import java.util.List;

public class LogItemadapter extends RecyclerView.Adapter<LogItemadapter.ViewHolder> {

    private List<LogItemView> mLogItemViewList;

    public LogItemadapter(List<LogItemView> updateList) {
        mLogItemViewList = updateList;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView logMessage;
        RecyclerView updateItemCardList;

        ViewHolder(View view) {
            super(view);
            logMessage = view.findViewById(R.id.log_message_textView);
            updateItemCardList = view.findViewById(R.id.update_item_recycler_view);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(
                LayoutInflater.from(parent.getContext()).inflate(R.layout.cardview_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        LogItemView logItemView = mLogItemViewList.get(position);
        holder.logMessage.setText(logItemView.getLogMessage());
    }

    @Override
    public int getItemCount() {
        return mLogItemViewList.size();
    }
}

File: app/src/main/java/net/xzos/upgradeall/ui/viewmodels/adapters/LogItemAdapter.kt
package net.xzos.upgradeall.ui.viewmodels.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import net.xzos.upgradeall.ui.viewmodels.view.holder.LogRecyclerViewHolder
import net.xzos.upgradeall.utils.FileUtil
import java.util.*


class LogItemAdapter(mLogList: LiveData<List<String>>, owner: LifecycleOwner) : RecyclerView.Adapter<LogRecyclerViewHolder>() {
    private val mLogMessages = ArrayList<String>()

    init {
        mLogList.observe(owner, Observer { logList ->
            renewLogMessage(logList)
        })
    }

    private fun renewLogMessage(stringList: List<String>) {
        if (mLogMessages != stringList) {
            val startChangeIndex: Int
            val index = mLogMessages.size - 1
            if (index == -1 || index > stringList.size || mLogMessages[index] != stringList[index]) {
                startChangeIndex = 0
                mLogMessages.clear()
                for (logMessage in stringList)
                    mLogMessages.add(logMessage)
            } else {
                startChangeIndex = index + 1
                for (i in index + 1 until stringList.size)
                    mLogMessages.add(stringList[i])
            }
            notifyItemRangeChanged(startChangeIndex, mLogMessages.size)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LogRecyclerViewHolder {
        val holder = LogRecyclerViewHolder(
                LayoutInflater.from(parent.context).inflate(android.R.layout.simple_expandable_list_item_1, parent, false))
        val logTextView = holder.logTextView
        logTextView.setOnClickListener {
            val context = logTextView.context
            FileUtil.clipStringToClipboard(logTextView.text, context)
        }
        return holder
    }

    override fun onBindViewHolder(holder: LogRecyclerViewHolder, position: Int) {
        val logMessage = mLogMessages[position]
        holder.logTextView.text = logMessage
    }

    override fun getItemCount(): Int {
        return mLogMessages.size
    }
}
